File: set-ime.m
#import <Carbon/Carbon.h>

int main() {
  @autoreleasepool {
    //
    // Read ARGV[1]
    //
    NSArray<NSString*> *argv = [[NSProcessInfo processInfo] arguments];
    if ([argv count] != 2) { return 1; }
    NSString *desired_ime_name = argv[1];

    //
    // Lookup all available IMEs and find desired IME
    //
    TISInputSourceRef chosen_ime = nil;
    NSArray *ime_list = (NSArray*)TISCreateInputSourceList(nil, false);
    for (id elem in ime_list) {
      TISInputSourceRef ime = (TISInputSourceRef)elem;
      NSString *ime_name = TISGetInputSourceProperty(ime, kTISPropertyLocalizedName);

      if ([desired_ime_name isEqualToString:ime_name]) {
        chosen_ime = ime;
        break;
      }
    }

    //
    // Check if the lookup was successful
    //
    if (!chosen_ime) {
      fprintf(stderr, "Input source '%s' is not available\n", [desired_ime_name UTF8String]);
      return 1;
    }

    //
    // Change IME
    //
    const OSStatus err = TISSelectInputSource(chosen_ime);
    if (err) {
      fprintf(stderr, "Could not change input language (OSStatus = %i)\n", (int)err);
      return 1;
    }
  }
}

File: set-ime.swift
import Carbon

//
// Read ARGV[1]
//
let argv = CommandLine.arguments
if argv.count != 2 {
  print("usage: set-ime <name>");
  exit(1)
}
let desired_ime_name = argv[1]

//
// Lookup all available IMEs and find desired IME
//
var chosen_ime: TISInputSource? = nil
let ime_list = TISCreateInputSourceList(nil, false).takeRetainedValue() as! [TISInputSource]
for ime in ime_list {
  if let ime_name_ptr = TISGetInputSourceProperty(ime, kTISPropertyLocalizedName) {
    let ime_name = Unmanaged<AnyObject>.fromOpaque(ime_name_ptr).takeUnretainedValue() as! String
    if (desired_ime_name == ime_name) {
      chosen_ime = ime
      break
    }
  }
}

//
// Check if the lookup was successful
//
if chosen_ime == nil {
  print("Input source '\(desired_ime_name)' is not available")
  exit(1)
}

//
// Change IME
//
let err = TISSelectInputSource(chosen_ime)
if err != 0 {
  print("Could not change input language (OSStatus = \(err))")
  exit(1)
}
